Source/Core/VideoCommon/TextureDecoder_Common.cpp

```diff
@@ -434,55 +435,206 @@ static inline u32 DecodePixel_Paletted(u16 pixel, TLUTFormat tlutfmt)
   }
 }
 
+/*  General formula for computing texture offset
+    u16 sBlk = s / blockWidth;
+    u16 tBlk = t / blockHeight;
+    u16 widthBlks = (width / blockWidth) + 1;
+    u32 base = (tBlk * widthBlks + sBlk) * blockWidth * blockHeight;
+    u16 blkS = s & (blockWidth - 1);
+    u16 blkT =  t & (blockHeight - 1);
+    u32 blkOff = blkT * blockWidth + blkS;
+*/
+
+static u8 TexDecoder_LoadTexel4(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 3;
+  u16 tBlk = t >> 3;
+  u16 widthBlks = (imageWidth >> 3) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u16 blkS = s & 7;
+  u16 blkT = t & 7;
+  u32 blkOff = (blkT << 3) + blkS;
+
+  int rs = (blkOff & 1) ? 0 : 4;
+  u32 offset = base + (blkOff >> 1);
+
+  return (*(src + offset) >> rs) & 0xF;
+}
+
+static u8 TexDecoder_LoadTexel8(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 3;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 3) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u16 blkS = s & 7;
+  u16 blkT = t & 3;
+  u32 blkOff = (blkT << 3) + blkS;
+
+  return *(src + base + blkOff);
+}
+
+static u16 TexDecoder_LoadTexel16(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 2;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 2) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 4;
+  u16 blkS = s & 3;
+  u16 blkT = t & 3;
+  u32 blkOff = (blkT << 2) + blkS;
+
+  u32 offset = (base + blkOff) << 1;
+  return *(u16*)(src + offset);
+}
+
+static u32 TexDecoder_LoadTexel32(const u8* src, int s, int t, int imageWidth)
+{
+  u16 sBlk = s >> 2;
+  u16 tBlk = t >> 2;
+  u16 widthBlks = (imageWidth >> 2) + 1;
+  u32 base = (tBlk * widthBlks + sBlk) << 5;  // shift by 5 is correct
+  u16 blkS = s & 3;
+  u16 blkT = t & 3;
+  u32 blkOff = (blkT << 2) + blkS;
+
+  u32 offset = (base + blkOff) << 1;
+  const u32 valAR = *(u16*)(src + offset);
+  const u32 valGB = *(u16*)(src + offset + 32);
+
+  // Swizzle to RBGA
+  return (valAR >> 8) | (valGB << 8) | ((valAR & 0xff) << 24);
+}
+
+#pragma pack(push, 1)
+struct cmprTexelData
+{
+  u16 color1;
+  u16 color2;
+  u32 colorSel;
+};
+#pragma pack(pop)
+
+static cmprTexelData TexDecoder_LoadCmprTexel(const u8* src, int s, int t, int widthBlks)
+{
+  u32 sBlk = s >> 3;
+  u32 tBlk = t >> 3;
+
+  u32 base = (tBlk * widthBlks + sBlk) << 5;
+  u32 blkS = s & 4;
+  u32 blkT = t & 4;
+  u32 blkOff = (blkT << 2) + (blkS << 1);
+
+  u32 offset = base + blkOff;
+
+  cmprTexelData ret;
+  std::memcpy(&ret, src + offset, sizeof(ret));
+  return ret;
+}
+
+static constexpr int CacheHashBits = 10;
+
+static constexpr u32 cmprCacheSet(u32 color1, u32 color2)
+{
+  return (color1 * color2 * 0x9E3779B1) >> (32 - CacheHashBits);
+}
+
+static constexpr u32 cmprCacheTag(u32 color1, u32 color2)
+{
+  return color1 | (color2 << 16);
+}
+
+struct CMPRCacheEntry
+{
+  u32 tag = {};
+  u32 colors[4] = {};
+};
+
+static CMPRCacheEntry cmprCache[1 << CacheHashBits] = {};
```

Comment:
`s_cmpr_cache`.  IMO it would also be a good idea to add a comment noting that this cache is only an implementation detail for performance and not an actual requirement for implementing the texture format.